js/modules/k6/ws/ws_test.go

```diff
@@ -274,7 +275,38 @@ func TestSession(t *testing.T) {
 		`))
 		assert.NoError(t, err)
 	})
-	assertSessionMetricsEmitted(t, stats.GetBufferedSamples(samples), "", sr("WSBIN_URL/ws-echo"), 101, "")
+
+	serverCloseTests := []struct {
+		name     string
+		endpoint string
+	}{
+		{"server_close_ok", "/ws-echo"},
+		// Ensure we correctly handle invalid WS server
+		// implementations that close the connection prematurely
+		// without sending a close control frame first.
+		{"server_close_invalid", "/ws-close-invalid"},
+	}
+
+	for _, tc := range serverCloseTests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := common.RunString(rt, sr(fmt.Sprintf(`
+			let closed = false;
+			let res = ws.connect("WSBIN_URL%s", function(socket){
+				socket.on("open", function() {
+					socket.send("test");
+				})
+				socket.on("close", function() {
+					closed = true;
+				})
+			});
+			if (!closed) { throw new Error ("close event not fired"); }
```

Comment:
I was surprised this check actually works... I would expect `ws.connect()` to return right after the connection is established, since `res` will have the `101 Switching Protocols` response, and the WS handlers to be called "much" later. So I think this test and others like it should fail :)

It seems like we're blocking on `ws.connect()` for the entire duration of the WS connection. Should this be a bug?

Reply:
it's literally what the [documentation](https://docs.k6.io/docs/connect-url-params-func) says it will do. This is because ... well we do not have an event loop so there is no other way for you to handle the asynchronous nature of ws except for us to make a mini event loop just for ws and than wait until the connection closes :).

Reply:
Also you are not using `ws-close` anywhere, either remove it or add it here as an additional test

Reply:
D'oh! Should've RTFM...

Thanks, yeah, I assumed it was because of no global event loop. This should be straightforward to fix once that's in.

I pushed 2ac9618b reverting most of the WS URL updates, so now we're using all 4 variants. I don't think we can get rid of either of them for now, unfortunately.